File: Algorithm/028. Implement strStr().py
# -*- coding: utf-8 -*-
# @Author: LC
# @Date:   2016-04-01 13:53:29
# @Last modified by:   LC
# @Last Modified time: 2016-04-10 16:23:08
# @Email: liangchaowu5@gmail.com

class Solution(object):
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        if len(needle) == 0:
            return 0
        if len(haystack) == 0: 
            return -1
            
        hlen = len(haystack)
        nlen = len(needle)
        for i in range(hlen-nlen+1):
            if haystack[i:i+nlen] == needle:
                return i
        return -1
            
            
            
            
            

File: Algorithm/Java/028. Implement strStr().java
/**
* Author: LC
* Date:   2016-11-06 09:55:32
* Last modified by:   WuLC
* Last Modified time: 2016-11-06 12:42:14
* Email: liangchaowu5@gmail.com
*/


// naive solution
// O(mn) time, O(mn) space
public class Solution 
{
    public int strStr(String haystack, String needle) 
    {
        int haystackLength = haystack.length(), needleLength = needle.length();
        for (int i=0;i <= haystackLength - needleLength; i++)
        {
            if (haystack.substring(i, i + needleLength).equals(needle))  return i;
        }
        return -1;
    }
}

// naive solution
// O(mn) time, O(1) space
public class Solution 
{
    public int strStr(String haystack, String needle) 
    {
        int haystackLength = haystack.length(), needleLength = needle.length(), i=0, j=0, tmp;
        if (needleLength==0) return 0;
        while (i < haystackLength)
        {
            tmp = i;
            j = 0;
            while (tmp < haystackLength && j < needleLength)
            {
                if (haystack.charAt(tmp) == needle.charAt(j))
                {
                    tmp++; 
                    j++;
                }
                else break;
            }
            if (j==needleLength) return i;
            i++;
        }
        return -1;
    }
}

//another more concise solution
public class Solution 
{
    public int strStr(String haystack, String needle) 
    {
        for (int i=0;;i++)
        {
            for(int j =0;;j++)
            {
                if (j==needle.length()) return i;
                if (i+j==haystack.length()) return -1;
                if (haystack.charAt(i+j)!=needle.charAt(j)) break;
            }
        }
    }
}

// KMP solution
// http://www.ruanyifeng.com/blog/2013/05/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm.html